File: src/test/java/kitchenpos/products/application/tobe/ProductServiceTest.java
package kitchenpos.products.application.tobe;

import kitchenpos.products.tobe.domain.infra.InMemoryProductRepository;
import kitchenpos.products.tobe.domain.entity.Product;
import kitchenpos.products.tobe.domain.ProductService;
import kitchenpos.products.tobe.domain.vo.DisplayedName;
import kitchenpos.products.tobe.domain.vo.Price;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import java.util.List;
import java.util.Optional;

import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.assertj.core.api.AssertionsForInterfaceTypes.assertThat;

public class ProductServiceTest {
    @DisplayName("상품의 이름에는 비속어가 포함될 수 없다")
    @Test
    void changeWrongNameTest() throws Exception {
        Price price = new Price(10);
        DisplayedName displayedName = new DisplayedName("wrong-name");

        ProductService service = new ProductService(new InMemoryProductRepository());
        assertThatThrownBy(() -> service.register(displayedName, price))
                .isInstanceOf(IllegalArgumentException.class);
    }


    @DisplayName("상품의 이름을 변경할 수 있다")
    @Test
    void changeGetNameTest() throws Exception {
        Price price = new Price(10);
        DisplayedName displayedName = new DisplayedName("name");

        ProductService service = new ProductService(new InMemoryProductRepository());
        Product pd = service.register(displayedName, price);

        service.changeName(pd.getId(), "new-name");

        Optional<Product> updatedPd = service.findById(pd.getId());
        Product upd = updatedPd.get();
        assertThat(upd.getName().value()).isEqualTo("new-name");
    }

    @DisplayName("상품을 등록할 수 있다")
    @Test
    void registerNewTest() {
        Price price = new Price(10);
        DisplayedName displayedName = new DisplayedName("name");

        ProductService service = new ProductService(new InMemoryProductRepository());
        service.register(displayedName, price);

        List<Product> pdList = service.getList();
        assertThat(pdList).hasSize(1);
    }
}

File: src/test/java/kitchenpos/products/application/tobe/ProductServiceTest.kt
package kitchenpos.products.application.tobe

import kitchenpos.common.price
import kitchenpos.menus.tobe.application.MenuService
import kitchenpos.menus.tobe.domain.MenuGroupService
import kitchenpos.menus.tobe.domain.MenuRepository
import kitchenpos.menus.tobe.domain.findByIdOrNull
import kitchenpos.menus.tobe.dto.`in`.MenuCreateRequest
import kitchenpos.menus.tobe.dto.`in`.MenuProductCreateRequest
import kitchenpos.products.tobe.application.ProductService
import kitchenpos.products.tobe.domain.ProductRepository
import kitchenpos.products.tobe.dto.int.ProductCreateRequest
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import java.math.BigDecimal

@SpringBootTest
class ProductServiceTest {
    @Autowired
    private lateinit var menuGroupService: MenuGroupService

    @Autowired
    private lateinit var productService: ProductService

    @Autowired
    private lateinit var menuService: MenuService

    @Autowired
    private lateinit var menuRepository: MenuRepository

    @Autowired
    private lateinit var productRepository: ProductRepository

    @Test
    fun `상품 변경 시 해당 상품을 포함하는 메뉴 가격 정책에 맞지 않는 전시 중인 메뉴의 전시상태 INACTIVATE`() {
        //given
        val menuGroup = menuGroupService.createMenuGroup("주 메뉴")
        val product = productService.createProduct(
            ProductCreateRequest(
                displayedName = "양념치킨",
                price = BigDecimal.valueOf(20000).price()
            )
        )
        val menu = menuService.createMenu(
            MenuCreateRequest(
                name = "치킨",
                price = BigDecimal.valueOf(15000).price(),
                displayed = true,
                menuGroupId = menuGroup.id,
                menuProducts = listOf(
                    MenuProductCreateRequest(
                        productId = product.id,
                        quantity = 1
                    )
                )
            )
        )

        //when
        val `만원` = BigDecimal.valueOf(10000).price()
        productService.changePrice(product.id, `만원`)

        //then
        assertThat(productRepository.findById(product.id)?.price?.value?.compareTo(`만원`.value)).isEqualTo(0)
        assertThat(menuRepository.findByIdOrNull(menu.id)?.displayStatus).isEqualTo(false)
    }
}
